Implement Label3D node.

bruvzg opened the pull request:
Implements https://github.com/godotengine/godot-proposals/issues/3219

<img width="597" alt="Screenshot 2022-04-20 at 10 29 17" src="https://user-images.githubusercontent.com/7645683/164174338-a0f84d7e-18a3-46ef-87ef-d0ea34363038.png">

- Implements `Label3D` node, with most of the `Label` features (except overrun, and visible characters).
- Adds `generate_mipmap` font import option.
- Adds some missing flags to the `Sprite3D`.

A minor convenience for normal fonts, but required to properly render MSDF fonts at arbitrary scale without 2D-in-3D viewport resolution restrictions.

<details>
  <summary>MSDF font zoom - click to expand</summary>

https://user-images.githubusercontent.com/7645683/164174441-83d8822d-6c25-4746-ae87-ba31f7add87b.mov

</details>

bruvzg committed:
Implement Label3D node.

Add "generate_mipmap" font import option.
Add some missing features to the Sprite3D.
Move BiDi override code from Control to TextServer.
Add functions to access TextServer font cache textures.
Add MSDF related flags and shader to the standard material.
Change standard material cache to use HashMap instead of Vector.

timothyqiu commented on editor/plugins/ot_features_plugin.cpp:
```diff
@@ -116,7 +118,19 @@ void OpenTypeFeaturesAdd::setup(Object *p_object) {
+	} else {
+		Label3D *l3d = Object::cast_to<Label3D>(edited_object);
+		if (l3d != nullptr) {
+			features = l3d->get_font()->get_feature_list();
```
This line crashes when I create a new `Label3D`, the new label has a null `Font`.

timothyqiu commented:
* If the `Font`'s data changes, i.e. replaced with another ttf file, the `Label3D` is not updated.
* if I reimport the font in-use, `Label3D` won't show anything until I reopen the scene.
* Clearing `Font` property crashes the editor.

bruvzg commented:
> This line crashes when I create a new Label3D, the new label has a null Font.
> If the Font's data changes, i.e. replaced with another ttf file, the Label3D is not updated.
> if I reimport the font in-use, Label3D won't show anything until I reopen the scene.
> Clearing Font property crashes the editor.

All should be fixed now.

Chaosus commented:
And it's possible to control a "depth" (volume on z axis) for the symbols in that label? If not, that would be interesting to add.

bruvzg commented:
> And it's possible to control a "depth" (volume on z axis) for the symbols in that label? If not, that would be interesting to add.

No, with this approach, each glyph is a flat quad with a character texture, so there's no depth.

There's a way to get Bézier shapes from the font (if it's a dynamic font), but they should be converted to polygons, triangulated and "extruded" to form a mesh. And last time I have checked, the triangulation algorithm included in the Godot wasn't working well on polygons with holes, so it might need much more work. Also, intersections of the chars probably will need some care to seamlessly connect.

I was thinking about implementing this as well, but probably as a separate `TextMesh`.

Calinou commented:
With the latest revision of this PR, creating a Label3D for the first time results in the following error message:

>   scene/3d/label_3d.cpp:423 - Condition "font.is_null()" is true.
Create Node

I think it would be good if the default project DynamicFont was used when no font is defined. This would allow for prototypes and level designers to quickly use Label3D for labeling various areas in levels without having to specify a font.

reduz commented:
@bruvzg You are just too awesome.

bruvzg commented:
- Added fallback to default project theme font, if no `Font` (or font with no `FontData`) is set.
- Now it's drawing invalid chars (as a solid rect, without hex-code texture).
- Added alpha cut modes (same as `Sprite3D`), alpha scissor is working fine with MSDF, but not so much with anti-aliased non MSDF fonts (needs `alpha_scissor_threshold` adjustment).
- `Material Override` and `Material Overlay` properties are hidden.
- Added auto wrap / fill align support.

bruvzg commented:
> Example with MSDF fonts (Godot icon is broken here, but font is OK):

That's a common issue most SVG to font / MSDF tools have, and related to the multiple overlapping color layers in the source icon. Converting source SVG to the single monochrome path usually helps:

```svg
<svg width="1024" height="1024" xmlns="http://www.w3.org/2000/svg"><path style="fill:#478cbf;fill-opacity:1;fill-rule:nonzero;stroke:none;stroke-width:1.33333" d="M417.615 76.875c-42.392 9.424-84.326 22.545-123.642 42.334.899 34.716 3.143 67.98 7.693 101.768-15.268 9.782-31.315 18.177-45.576 29.628-14.49 11.148-29.29 21.813-42.41 34.85-26.212-17.337-53.955-33.63-82.535-48.012C100.337 270.6 71.53 306.383 48 346.428c18.493 29.029 38.33 58.205 56.7 80.959v245.765c.449.004.898.021 1.343.063l150.67 14.527a16.223 16.223 0 0 1 14.627 15.024l4.646 66.51 131.43 9.376 9.055-61.384a16.222 16.222 0 0 1 16.05-13.858h158.961c8.047 0 14.873 5.899 16.047 13.858l9.055 61.384 131.434-9.377 4.642-66.51a16.23 16.23 0 0 1 14.627-15.023l150.611-14.527c.446-.042.89-.059 1.34-.063v-19.611l.063-.02V427.387c21.216-26.71 41.307-56.172 56.699-80.96-23.523-40.044-52.345-75.828-83.152-108.984-28.573 14.382-56.325 30.675-82.537 48.012-13.117-13.037-27.89-23.702-42.4-34.85-14.258-11.45-30.324-19.846-45.563-29.628 4.537-33.788 6.78-67.052 7.683-101.768-39.32-19.79-81.249-32.91-123.662-42.334-16.933 28.46-32.42 59.28-45.906 89.408-15.993-2.672-32.06-3.662-48.149-3.853v-.026c-.112 0-.216.026-.312.026-.1 0-.205-.026-.305-.026v.026c-16.117.191-32.17 1.18-48.168 3.853-13.478-30.129-28.955-60.948-45.914-89.408zM298.416 436.398c50.151 0 90.799 40.618 90.799 90.752 0 50.168-40.648 90.809-90.799 90.809-50.126 0-90.787-40.64-90.787-90.809 0-50.134 40.66-90.752 90.787-90.752zm427.178 0c50.122 0 90.779 40.618 90.779 90.752 0 50.168-40.657 90.809-90.78 90.809-50.159 0-90.806-40.64-90.806-90.809 0-50.134 40.647-90.752 90.807-90.752zm-418.498 35.87c-33.285 0-60.27 26.993-60.27 60.27 0 33.275 26.985 60.245 60.27 60.245 33.3 0 60.271-26.97 60.271-60.246s-26.97-60.27-60.271-60.27zm409.78 0c-33.275 0-60.235 26.993-60.235 60.27 0 33.275 26.96 60.245 60.236 60.245 33.31 0 60.271-26.97 60.271-60.246s-26.962-60.27-60.271-60.27zm-204.882 17.24c16.143 0 29.254 11.908 29.254 26.56v83.59c0 14.665-13.111 26.563-29.254 26.563-16.142 0-29.226-11.898-29.226-26.563v-83.59c0-14.652 13.084-26.56 29.226-26.56zM104.451 705.66c.063 14.561.248 30.514.248 33.69 0 143.085 181.512 211.86 407.026 212.65h.553c225.513-.79 406.962-69.565 406.962-212.65 0-3.235.195-19.12.262-33.69l-135.43 13.063-4.668 66.865c-.562 8.059-6.972 14.472-15.03 15.05l-160.49 11.452c-.39.029-.782.043-1.17.043-7.975 0-14.856-5.853-16.034-13.862l-9.203-62.416H446.525l-9.203 62.416c-1.236 8.4-8.746 14.44-17.205 13.819L259.63 800.639c-8.059-.579-14.47-6.992-15.031-15.051l-4.666-66.865-135.48-13.063z"/></svg>
```

![Screenshot 2022-04-21 at 11 00 36](https://user-images.githubusercontent.com/7645683/164408683-321750d9-325b-488f-b772-3f63958832f6.png)

Calinou commented:
To prevent distant DynamicFonts from getting grainy, would it be possible to have a Mipmaps toggle?

albinaask commented:
Looks awsome! I'd perhaps suggest a bill board option to always show the text towards you but to stay in place. It would be useful for waypoints, area markers, signs etc?

If I recall correctly it can be done through the transformation matrix somehow... did it ages ago for a particle system i think.

Calinou commented:
> Looks awsome! I'd perhaps suggest a bill board option to always show the text towards you but to stay in place. It would be useful for waypoints, area markers, signs etc?

What's the difference with the current billboard modes proposed? There's already Billboard and Y-Billboard.

bruvzg commented:
> To resolve this issue, could the outline vertices be slightly offset from the main font when the Alpha Cut property is Discard? The outline would then be drawn behind the main font. Using an offset of pixel_size units should work well enough most of the time.

This will help only at some distances and create visible gap, `ALPHA_CUT_DISABLED` is setting render priority to solve this, other modes seems to ignore it. But added it anyway (only for non `ALPHA_CUT_DISABLED` modes).

<img width="579" alt="Screenshot 2022-04-22 at 09 05 40" src="https://user-images.githubusercontent.com/7645683/164613702-0b0d0463-cd76-4f5b-bdc6-6aede7a193d0.png">

> Alpha Scissor works as expected, but the default value (0.98) seems pretty high:

Changed default value to `0.5`.

> Normals seem to be inverted on the generated mesh. When Shaded is enabled, faces that should be bright are dark and vice versa

I have completely missed setting normals and tangents at all, should be fixed now.

https://user-images.githubusercontent.com/7645683/164614687-e54507d5-111e-43c7-84fb-13a1a2e4ebb2.mov

> Long texts can cause mesh generation to fail with the following error message.

Now it should create new surface only if a new material is used (one surface for each of the font cache textures used).

> I'd recommend overriding GeometryInstance3D's Cast Shadow property to Off by default, so that Label3D doesn't cast shadows by default. This should improve performance a little by excluding the geometry from shadow passes, resulting in fewer draw calls.

Done.

> The Width property should be hidden if Autowrap Mode is set to Off. The value should still be Off by default, but I suggest making the default width a value like 500, and set a property hint to disallow entering values below 1. With the default font size and a width of 500, arbitrary wrapping looks like this

`width` is also used for the `fill` alignment.

bruvzg commented:
> Would it be possible to expose the texture filtering mode as is done in BaseMaterial3D? This would make it possible to enable mipmaps

Added `generate_mipmaps` option to the `Font` import settings (disabled by default). And `texture_filter` flag to the `Label3D` and `Sprite3D`.

*Left to right: Nearest, Linear, Linear Mipmap, Linear Mipmap Anisotropic.*
<img width="801" alt="Screenshot 2022-04-22 at 10 40 43" src="https://user-images.githubusercontent.com/7645683/164641679-af75b8ec-853c-412b-b6b1-3e439b7a17cb.png">

Note: to have text with sharp pixelated edges when using a dynamic font, `aa` should be disabled in both `Label3D` texture filter (filter applied during `Label3D` rendering) and `Font` import settings (filter applied internally by freetype during rasterization of the glyph).

<img width="801" alt="Screenshot 2022-04-22 at 10 46 35" src="https://user-images.githubusercontent.com/7645683/164642987-664de1dc-0c10-40c9-8ea7-606ccc492161.png">

fire reviewed (approved):
I made a label and aside from not having a default font, the Label3D worked.

Calinou reviewed (approved):
I tested the latest revision of this PR, everything seems to be working great :slightly_smiling_face: 

![image](https://user-images.githubusercontent.com/180032/164731795-e8482a21-3ed2-402c-ac2c-d4115722b818.png)

**Testing project:** [test_label_3d_2.zip](https://github.com/godotengine/godot/files/8541682/test_label_3d_2.zip)

bruvzg commented:
>>    And it's possible to control a "depth" (volume on z axis) for the symbols in that label? If not, that would be interesting to add.
>
>No, with this approach, each glyph is a flat quad with a character texture, so there's no depth.
>
>There's a way to get Bézier shapes from the font (if it's a dynamic font), but they should be converted to polygons, triangulated and "extruded" to form a mesh. And last time I have checked, the triangulation algorithm included in the Godot wasn't working well on polygons with holes, so it might need much more work. Also, intersections of the chars probably will need some care to seamlessly connect.
>
> I was thinking about implementing this as well, but probably as a separate `TextMesh`.

I think I got glyph outlines convex decomposition / triangulation working (at least with the proper fonts, without self-intersection), so the most complex part is done. Still needs a bit of work to get rid of glyph-to-glyph intersections, but from this point, mesh generation should be simple.

![Screenshot 2022-04-25 at 11 35 30](https://user-images.githubusercontent.com/7645683/165052040-7f53527e-6a0b-4ffc-ab00-9797fcdb9ec6.png)

Probably will make a second PR for the `TextMesh` later this week.

bruvzg commented:
> Probably will make a second PR for the TextMesh later this week.

Text mesh generation PR draft - https://github.com/godotengine/godot/pull/60507

akien-mga reviewed (approved):
Looks good, and amazing feature!

akien-mga commented:
Thanks!

golddotasksquestions commented:
Wow, this is really an incredible new feature! I just gave it a quick test and there is really nothing I could think of left desired! Huge thanks to @bruvzg (also to @Calinou for the continuous testing) 
Everything just works super easy out of the box and even the most important Inspector properties (like Font and Size) are right there at the top of the Inspector where you would expect them from a general user perspective, super easy and intuitive to find and change.

May I ask what the reason was for limiting the font size to 127?

I'm not sure where to ask for this, because this seems like too small of a request to open a proposal for, but would it be possible to have the "Font", "Font Size", "Font Outline" properties of the Label3D also at the same location in the Inspector for the Label2D?
![image](https://user-images.githubusercontent.com/47016402/165933114-7b09dee2-a186-4b81-9593-08f610fee893.png)

Calinou commented:
> The limit is copy-pasted from the importer settings, probably should be changed to 127,or_greater in both cases (like a normal Label it should not have any issues with the font sizes at least up to 4K).

This might cause a freeze if the user enters a very large value like `1000`.

golddotasksquestions commented:
> For rasterized DynamicFonts, this might cause a freeze if the user enters a very large value like 1000. 

Because Godot has to render (rasterize) each Glyph of the font? Tbh I'd rather be able to have to wait a little and have a progress bar pop up (with the rest of the Editor tinted dark so it's clear it's inactive), than not being able to choose larger (= higher res) fonts sizes.

Calinou commented:
> Because Godot has to render (rasterize) each Glyph of the font?

Yes.

> Tbh I'd rather be able to have to wait a little and have a progress bar pop up (with the rest of the Editor tinted dark so it's clear it's inactive), than not being able to choose larger (= higher res) fonts sizes.

The issue is that this can cause stuttering during gameplay, especially on slower CPUs (font rasterization is done by FreeType on the CPU). This is especially noticeable on mobile devices, and it can result in difficult-to-diagnose slowdowns.

golddotasksquestions commented:
> The issue is that this can cause stuttering during gameplay,

I don't think this is an issue at all. Setting font size is typically an editor usecase. All it needs is proper documentation which mentions

"If you set the fontsize during runtime, Godot needs to render (rasterize) the every glyph of the font. For large font sizes, this will lead to a freeze until the work is done. Consider using separate threads when you want to set large font sized during gameplay."

Calinou commented:
> I don't think this is an issue at all. Setting font size is typically an editor usecase. All it needs is proper documentation which mentions something like:

What I mean is that if you use a rasterized DynamicFont, each glyph will be rasterized the first time it is used. During gameplay, if a string changes and requires many new glyphs to be rasterized, this can take a long time with large font sizes.
